File: data/Algorithms/Binary Tree.c
#include<stdio.h>
#include<conio.h>
#include<stdlib.h>
#include<malloc.h>

struct node
{
    char info;
    node *left, *right;
}*root,*ptr;

void inorder(node *tree)
{
    if(tree != NULL)
    {
        inorder(tree->left);
        printf("%c  ", tree->info);
        inorder(tree->right);
    }
}

void preorder(node *tree)
{
    if(tree != NULL)
    {
        printf("%c  ", tree->info);
        preorder(tree->left);
        preorder(tree->right);
    }
}

void postorder(node *tree)
{
    if(tree != NULL)
    {
        postorder(tree->left);
        postorder(tree->right);
        printf("%c  ", tree->info);
    }
}

node *create(char x)
{
    node *n = (node*)malloc(sizeof(node));
    n->info = x;
    n->left = NULL;
    n->right = NULL;
    return n;
}

void setleft(node *tree, char x)
{
    if(tree ==NULL) printf("Void Insertion");
    else if(tree->left != NULL) printf("Invalid Insertion");
    else tree->left = create(x);
}

void setright(node *tree, char x)
{
    if(tree ==NULL) printf("Void Insertion");
    else if(tree->right != NULL) printf("Invalid Insertion");
    else tree->right = create(x);
}

int main()
{
    int choice;
    char data;
    printf("Binary Tree");
    printf("\nRoot Element? ");
    fflush(stdin);
    scanf("%c", &data);
    root = create(data);
    ptr = root;			
    while(1)
    {
        system("CLS");
        printf("Inorder\n");
        inorder(root);
        printf("\nPreorder\n");
        preorder(root);
        printf("\nPostorder\n");
        postorder(root);

        printf("\nBinary Tree\n");
        printf("Press 1 to insert at left\n");
        printf("Press 2 to insert at right\n");
        printf("Press 3 to traverse to root\n");
        printf("Press 4 to traverse left\n");		
        printf("Press 5 to traverse right\n");
        printf("Press 0 to exit\n");
        printf("Choice? ");
        scanf("%d", &choice);
        switch(choice)
        {
            case  1 : printf("Element? ");
                    fflush(stdin);
                    scanf("%c", &data);
                    setleft(ptr, data);
                    break;
            case  2 : printf("Element? ");
                    fflush(stdin);
                    scanf("%c", &data);
                    setright(ptr, data);
                    break;
            case  3 : ptr = root;
                    break;
            case  4 : ptr = ptr->left;
                    break;
            case  5 : ptr = ptr->right;
                    break;
            case  0 : exit(0);
            default : printf("Wrong Choice");			
        }
    }
    return 0;
}

File: data/Algorithms/BinaryTree.py
class Node:
    
    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

    def PrintTree(self):
        if self.left:
            self.left.PrintTree()
        print(self.data),
        if self.right:
            self.right.PrintTree()

    def inorderTraversal(self, node):
        res = []
        if node:
            res = self.inorderTraversal(node.left)
            res.append(node.data)
            res = res + self.inorderTraversal(node.right)
        return res

    def preorderTraversal(self, node):
        res = []
        if node:
            res.append(node.data)
            res = res + self.preorderTraversal(node.left)
            res = res + self.preorderTraversal(node.right)
        return res
    
    def postorderTraversal(self, node):
        res = []
        if node:
            res = self.postorderTraversal(node.left)
            res = res + self.postorderTraversal(node.right)
            res.append(node.data)
        return res
    
    def __str__(self):
        return f"{self.data}"

if __name__ == "__main__":
    root = Node(27)
    root.left = Node(14)
    root.right = Node(35)
    root.left.left = Node(10)
    root.left.right = Node(19)
    root.right.left = Node(31)
    root.right.right = Node(42)
    print("Inorder:")
    print(*root.inorderTraversal(root))
    print("Preorder:")
    print(*root.preorderTraversal(root))
    print("Postorder:")
    print(*root.postorderTraversal(root))
